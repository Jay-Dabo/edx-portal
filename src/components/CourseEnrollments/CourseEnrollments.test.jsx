import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import { Pagination } from '@edx/paragon';

import { CourseEnrollments } from './index';
import {
  mockCourseEnrollments,
  COURSE_ENROLLMENTS_COUNT,
} from './CourseEnrollments.mocks';

const CourseEnrollmentsWrapper = props => (
  <MemoryRouter>
    <CourseEnrollments
      getCourseEnrollments={() => {}}
      {...props}
    />
  </MemoryRouter>
);

describe('<CourseEnrollments />', () => {
  let wrapper;

  describe('renders correctly', () => {
    it('calls getCourseEnrollments prop', () => {
      const getCourseEnrollmentsMock = jest.fn();
      mount((
        CourseEnrollmentsWrapper({
          getCourseEnrollments: getCourseEnrollmentsMock,
        })
      ));
      expect(getCourseEnrollmentsMock).toHaveBeenCalled();
    });

    it('with enrollment data', () => {
      const tree = renderer
        .create((
          CourseEnrollmentsWrapper({
            enrollments: mockCourseEnrollments,
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with error state', () => {
      const tree = renderer
        .create((
          CourseEnrollmentsWrapper({
            error: Error('Network Error'),
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with loading state', () => {
      const tree = renderer
        .create((
          CourseEnrollmentsWrapper({
            loading: true,
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('formatEnrollmentData', () => {
    const expectedData = [{
      ...mockCourseEnrollments.results[0],
      course_start: 'September 1, 2016',
      course_end: 'December 1, 2016',
      has_passed: 'Yes',
      last_activity_date: 'June 23, 2017',
      passed_timestamp: 'May 9, 2017',
      enrollment_created_timestamp: 'June 27, 2014',
      user_account_creation_timestamp: 'February 12, 2015',
    }, {
      ...mockCourseEnrollments.results[1],
      course_start: 'September 1, 2016',
      course_end: 'December 1, 2016',
      has_passed: 'No',
      last_activity_date: 'June 23, 2017',
      passed_timestamp: null,
      enrollment_created_timestamp: 'June 27, 2014',
      user_account_creation_timestamp: 'February 12, 2015',
    }];

    beforeEach(() => {
      wrapper = shallow((
        <MemoryRouter>
          <CourseEnrollments
            getCourseEnrollments={() => {}}
            enrollments={mockCourseEnrollments}
          />
        </MemoryRouter>
      )).find(CourseEnrollments);
    });

    it('handles empty enrollments data', () => {
      wrapper.dive().setProps({
        enrollments: {},
      }, () => {
        expect(wrapper.dive().state('enrollments')).toHaveLength(0);
      });
    });

    it('handles non-empty enrollments data', () => {
      const enrollmentsData = wrapper.dive().state('enrollments');

      expect(enrollmentsData).toHaveLength(COURSE_ENROLLMENTS_COUNT);
      expect(enrollmentsData[0]).toEqual(expectedData[0]);
      expect(enrollmentsData[1]).toEqual(expectedData[1]);
    });

    it('does not override state enrollments when props enrollments changes with existing enrollments', () => {
      const currentEnrollments = wrapper.dive().state('enrollments');
      expect(currentEnrollments[0]).toEqual(expectedData[0]);
      expect(currentEnrollments[1]).toEqual(expectedData[1]);

      wrapper.dive().setProps({
        enrollments: mockCourseEnrollments,
      }, () => {
        const updatedEnrollments = wrapper.dive().state('enrollments');
        expect(updatedEnrollments[0]).toEqual(expectedData[0]);
        expect(updatedEnrollments[1]).toEqual(expectedData[1]);
      });
    });
  });

  it('handles pagination correctly', () => {
    const mockHistoryPush = jest.fn();
    wrapper = mount((
      <CourseEnrollments
        getCourseEnrollments={() => {}}
        enrollments={mockCourseEnrollments}
        history={{ push: mockHistoryPush }}
      />
    ));

    const pagination = wrapper.find(CourseEnrollments).find(Pagination);
    pagination.find('button').last().simulate('click'); // click `Next` button
    expect(mockHistoryPush).toHaveBeenCalledWith('?page=2');
    expect(wrapper.find('CourseEnrollments').instance().state.currentPage).toEqual(2);
  });
});
