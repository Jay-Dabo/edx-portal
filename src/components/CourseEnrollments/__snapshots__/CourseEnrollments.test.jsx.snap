// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseEnrollments /> renders correctly with enrollment data 1`] = `
<div>
  <div
    className={undefined}
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="table-responsive"
        >
          <table
            className="table table-sm table-striped"
          >
            <thead
              className=""
            >
              <tr
                className=""
              >
                <th
                  className=""
                  scope="col"
                >
                  Email
                </th>
                <th
                  className=""
                  scope="col"
                >
                  Course
                </th>
                <th
                  className=""
                  scope="col"
                >
                  Course End
                </th>
                <th
                  className=""
                  scope="col"
                >
                  Mode
                </th>
                <th
                  className=""
                  scope="col"
                >
                  Passed
                </th>
                <th
                  className=""
                  scope="col"
                >
                  Last Activity Date
                </th>
              </tr>
            </thead>
            <tbody
              className=""
            >
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  No
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
              <tr
                className=""
              >
                <td
                  className=""
                >
                  test@example.com
                </td>
                <td
                  className=""
                >
                  All about acceptance testing!
                </td>
                <td
                  className=""
                >
                  December 1, 2016
                </td>
                <td
                  className=""
                >
                  verified
                </td>
                <td
                  className=""
                >
                  Yes
                </td>
                <td
                  className=""
                >
                  June 23, 2017
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col d-flex justify-content-center"
      >
        <nav
          aria-label="Course Enrollments Pagination"
          className=""
        >
          <div
            aria-atomic={true}
            aria-live="polite"
            aria-relevant="text"
            className="sr-only"
          >
            Page 1, Current Page, of 2
          </div>
          <ul
            className="pagination"
          >
            <li
              className="page-item disabled"
            >
              <button
                aria-label="Previous"
                className="btn previous page-link"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                tabIndex="-1"
                type="button"
              >
                <div>
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-chevron-left mr-2"
                      id="pagination-2"
                    />
                  </div>
                  Previous
                </div>
              </button>
            </li>
            <li
              className="page-item"
            >
              <button
                aria-label="Next, Page 2"
                className="btn next page-link"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                tabIndex={undefined}
                type="button"
              >
                <div>
                  Next
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-chevron-right ml-2"
                      id="pagination-3"
                    />
                  </div>
                </div>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CourseEnrollments /> renders correctly with error state 1`] = `
<div>
  <div
    className="alert fade alert-danger show"
    hidden={false}
    role="alert"
  >
    <div
      className="alert-dialog"
    >
      There was a problem fetching course enrollments: Network Error
    </div>
  </div>
</div>
`;

exports[`<CourseEnrollments /> renders correctly with loading state 1`] = `
<div>
  <div
    className="loading"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;
